### PR TITLE
Add dbaas resources

### PR DESCRIPTION
**Pull request by schirevko**

### CODE REVIEW

File: mcs/data_source_mcs_db_database.go
Comment (irlndts) [inline mcs/data_source_mcs_db_database.go:52]: а если id будет a/b/c, такое возможно? в массиве будет "a" и "b/c", это правильно будет работать с условием в следующей строке?

Comment (schirevko) [reply mcs/data_source_mcs_db_database.go:52]: Условие проверяет, что разделение на два элемента произошло, оно и произойдет как ты написал - "a" и "b/c"
В массиве будет два элемента - "a" и "b/c", следующее условие и проверит, что именно два элемента (эти) и будут в массиве, отработает верно.

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [inline docs/resources/mcs_db_cluster.md:22]: align spaces in this block, pls

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [inline docs/resources/mcs_db_cluster.md:59]: do not mention ipv6 here and everywhere

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [inline docs/resources/mcs_db_cluster.md:49]: move this up to name to make structure of both cluster and cluster with shards more similar

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [inline docs/resources/mcs_db_cluster.md:18]: sort keys in this block to have the same order with reference

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [reply docs/resources/mcs_db_cluster.md:22]: or do not add align spaces everywhere, in other examples as well

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [inline docs/resources/mcs_db_cluster.md:57]: The port id of the network. Add id to other such places as well

File: docs/resources/mcs_db_user.md
Comment (ftersin) [inline docs/resources/mcs_db_user.md:33]: it's shame we propose users to have password in tf files. should we release this?

File: examples/create-mcs-db-cluster-with-shards/provider.tf
Comment (ftersin) [inline examples/create-mcs-db-cluster-with-shards/provider.tf:12]: how does mcs provider work w/o this?

File: examples/create-mcs-db-database/main.tf
Comment (ftersin) [inline examples/create-mcs-db-database/main.tf:26]: i think we should not use az-related volume types in docs since they are deprecated

File: examples/create-mcs-db-database/main.tf
Comment (ftersin) [inline examples/create-mcs-db-database/main.tf:22]: please sort keys to the order of reference

File: docs/resources/mcs_db_cluster.md
Comment (sti-jans) [reply docs/resources/mcs_db_cluster.md:22]: you can use terraform fmt for examples

File: docs/resources/mcs_db_user.md
Comment (ftersin) [reply docs/resources/mcs_db_user.md:33]: can be resolved by terraform itself https://learn.hashicorp.com/tutorials/terraform/sensitive-variables

File: examples/create-mcs-db-cluster-with-shards/provider.tf
Comment (irlndts) [inline examples/create-mcs-db-cluster-with-shards/provider.tf:5]: I've checked, we support default value in code and it's the same as provided.
I think it's better not to add it to examples just not to disturb clients here.
Please remove it here and below.

I'm removing them from old examples as well.
https://github.com/MailRuCloudSolutions/terraform-provider-mcs/pull/21

We'll add the description about this optional value to docs then (you don't need to do this).

File: mcs/resource_mcs_db_user.go
Comment (anuriq) [inline mcs/resource_mcs_db_user.go:20]: extract to constants

File: mcs/resource_mcs_db_instance.go
Comment (anuriq) [inline mcs/resource_mcs_db_instance.go:15]: could declare as a time type from the beginning

File: mcs/resource_mcs_db_instance.go
Comment (anuriq) [inline mcs/resource_mcs_db_instance.go:295]: in public API it's called `allow_remote_access`. is it a good idea to change it in TF?

File: mcs/db_cluster_client.go
Comment (anuriq) [inline mcs/db_cluster_client.go:280]: All GET requests should have retries. For examples, get /instance/.../databases - may be successful after several retries.

File: mcs/db_instance_client.go
Comment (anuriq) [inline mcs/db_instance_client.go:652]: ALL delete requests should have retries for non successful code

**Review by anuriq — CHANGES_REQUESTED** [review]

File: mcs/resource_mcs_db_instance.go
Comment (ftersin) [reply mcs/resource_mcs_db_instance.go:295]: The question is for Lazarenko and Rann. The reference tf provided by Dmitry contains public_access exactly.

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md:18]: Fixed

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

File: examples/create-mcs-db-cluster-with-shards/provider.tf
Comment (schirevko) [reply examples/create-mcs-db-cluster-with-shards/provider.tf]: The same field (user_domain_name) is not present in working examples for kubernetes cluster

File: examples/create-mcs-db-database/main.tf
Comment (schirevko) [reply examples/create-mcs-db-database/main.tf]: Fixed

File: examples/create-mcs-db-database/main.tf
Comment (schirevko) [reply examples/create-mcs-db-database/main.tf:22]: Fixed

File: mcs/db_database.go
Comment (schirevko) [reply mcs/db_database.go]: Changed return to `return exists, err`

File: mcs/resource_mcs_db_instance.go
Comment (schirevko) [reply mcs/resource_mcs_db_instance.go]: Fixed

File: mcs/resource_mcs_db_user_test.go
Comment (schirevko) [reply mcs/resource_mcs_db_user_test.go:162]: Why not? This function will be run after terraform destroys resource for testing purposes, and if user is still existing, we should indicate it by returning error.

File: examples/create-mcs-db-cluster-with-shards/provider.tf
Comment (schirevko) [reply examples/create-mcs-db-cluster-with-shards/provider.tf]: Fixed

File: docs/resources/mcs_db_cluster_with_shards.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster_with_shards.md]: Fixed

File: mcs/db_cluster_client.go
Comment (schirevko) [reply mcs/db_cluster_client.go:9]: We followed kubernetes_cluster_client style and it has empty docstrings too: [https://github.com/MailRuCloudSolutions/terraform-provider-mcs/blob/f154f5d3e75b71dfde21a7ed5e8fa26ba2ab65a2/mcs/kubernetes_cluster_client.go#L62](https://github.com/MailRuCloudSolutions/terraform-provider-mcs/blob/f154f5d3e75b71dfde21a7ed5e8fa26ba2ab65a2/mcs/kubernetes_cluster_client.go#L62)

File: mcs/db_instance.go
Comment (schirevko) [reply mcs/db_instance.go:19]: Fixed

File: mcs/resource_mcs_db_cluster.go
Comment (schirevko) [reply mcs/resource_mcs_db_cluster.go]: Fixed

File: mcs/resource_mcs_db_user.go
Comment (schirevko) [reply mcs/resource_mcs_db_user.go]: Fixed

File: mcs/resource_mcs_db_instance.go
Comment (schirevko) [reply mcs/resource_mcs_db_instance.go]: We followed kubernetes_cluster style, and it also declares these constants as integers:
https://github.com/MailRuCloudSolutions/terraform-provider-mcs/blob/f154f5d3e75b71dfde21a7ed5e8fa26ba2ab65a2/mcs/resource_mcs_kubernetes_cluster.go#L14

File: mcs/resource_mcs_db_user_test.go
Comment (sti-jans) [reply mcs/resource_mcs_db_user_test.go:162]: Yeap. I meant that we can return error in the first match.
Anyway, this is not critical. You can leave everything as it is.

File: docs/resources/mcs_db_cluster.md
Comment (ftersin) [reply docs/resources/mcs_db_cluster.md]: Now the order looks better but still not consistent.
1 Orders of keys in instance and cluster differs from each other
2 Keys related to VM need to be grouped (flavor, AZ, keypair, public access)
3 So do volume related keys (size, type, autoexpand, wal)

Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

File: mcs/db_cluster_client.go
Comment (anuriq) [reply mcs/db_cluster_client.go:9]: How can this possibly be a code style? If somewhere code is bad, it doesn't mean, that we also should write a bad code. Please...

Comment (schirevko) [reply mcs/db_cluster_client.go]: Added docstrings

File: mcs/resource_mcs_db_instance.go
Comment (schirevko) [reply mcs/resource_mcs_db_instance.go]: Fixed

File: mcs/db_cluster_client.go
Comment (ftersin) [reply mcs/db_cluster_client.go:281]: Lets do it in a separate pr to make common solution inside mcs provider. FYI we are going to adopt OS provider for IaaS services during Q3, so this would be (and should be) a preparation task for that activity.

File: mcs/db_instance_client.go
Comment (ftersin) [reply mcs/db_instance_client.go:646]: The same, lets do it in a separate pr.

File: examples/create-mcs-db-cluster-with-shards/provider.tf
Comment (irlndts) [reply examples/create-mcs-db-cluster-with-shards/provider.tf]: thx

File: mcs/resource_mcs_db_instance.go
Comment (anuriq) [reply mcs/resource_mcs_db_instance.go:296]: Talked with them. Agreed on `floating_ip_enabled` - same as in Magnum

Comment (schirevko) [reply mcs/resource_mcs_db_instance.go]: Fixed

**Comment by irlndts** [comment]
rebase pls

**Comment by schirevko** [comment]
> rebase and `golangci-lint run ./...` pls

done

File: docs/resources/mcs_db_cluster.md
Comment (anuriq) [inline docs/resources/mcs_db_cluster.md:53]: Boolean field that indicates whether floating ip is created for cluster or instance.

**Comment by anuriq** [comment]
LGTM

File: docs/resources/mcs_db_cluster.md
Comment (schirevko) [reply docs/resources/mcs_db_cluster.md]: Fixed

**Review by irlndts — APPROVED** [review]

**Review by irlndts — APPROVED** [review]
I'm okay with the diff, but i'm here more about code-style not about the logic. 
Please wait for @anuriq to solve all the questions before merge.

**Review by anuriq — APPROVED** [review]

**Pull request merged**
